File: e2e-tests/page-objects/LoginPage.js
import BasePage from './BasePage.js';

class LoginPage extends BasePage {
    get pageTitle() { return $('h1'); }
    get emailInput() { return $('input#email'); }
    get loginButton() { return $('button[type="submit"]'); }
    get infoAlert() { return $('.MuiAlert-standardInfo'); }
    get errorAlert() { return $('.MuiAlert-standardError'); }

    async open() {
        await super.open('/login');
    }

    async getPageTitle() {
        const title = await this.pageTitle;
        await title.waitForDisplayed();
        return await title.getText();
    }

    async enterEmail(email) {
        const input = await this.emailInput;
        await input.waitForDisplayed();
        await input.clearValue();
        await input.setValue(email);
    }

    async clearEmailField() {
        const input = await this.emailInput;
        await input.clearValue();
        await browser.keys(['Tab']);
    }

    async clickLoginButton() {
        const button = await this.loginButton;
        await button.waitForDisplayed();
        await button.click();
    }

    async isLoginButtonEnabled() {
        const button = await this.loginButton;
        return await button.isEnabled();
    }

    async isLoginButtonDisabled() {
        const button = await this.loginButton;
        const isDisabled = await button.getAttribute('disabled');
        return isDisabled !== null;
    }

    async isEmailInputDisplayed() {
        const input = await this.emailInput;
        return await input.isDisplayed();
    }

    async isLoginButtonDisplayed() {
        const button = await this.loginButton;
        return await button.isDisplayed();
    }

    async isInfoMessageDisplayed() {
        const alert = await this.infoAlert;
        return await alert.isDisplayed();
    }

    async getErrorMessage() {
        const alert = await this.errorAlert;
        await alert.waitForDisplayed({ timeout: 5000 });
        return await alert.getText();
    }

    async isErrorDisplayed() {
        try {
            const alert = await this.errorAlert;
            return await alert.isDisplayed();
        } catch {
            return false;
        }
    }

    async login(email) {
        await this.enterEmail(email);
        await this.clickLoginButton();
    }
}

export default new LoginPage();

File: selenium-tests/src/main/java/com/timesheet/pages/LoginPage.java
package com.timesheet.pages;

import com.timesheet.utils.ConfigReader;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

public class LoginPage extends BasePage {

    private final By emailInput = By.id("email");
    private final By loginButton = By.cssSelector("button[type='submit']");
    private final By pageTitle = By.xpath("//h1[contains(text(), 'Time Tracker')]");
    private final By infoAlert = By.cssSelector(".MuiAlert-standardInfo");
    private final By errorAlert = By.cssSelector(".MuiAlert-standardError");
    private final By loadingSpinner = By.cssSelector(".MuiCircularProgress-root");

    public LoginPage(WebDriver driver) {
        super(driver);
    }

    public void navigateToLoginPage() {
        navigateTo(ConfigReader.getBaseUrl() + "/login");
        waitUtils.waitForElementVisible(emailInput);
    }

    public boolean isLoginPageDisplayed() {
        try {
            waitUtils.waitForElementVisible(pageTitle);
            return isElementDisplayed(pageTitle) && isElementDisplayed(emailInput);
        } catch (Exception e) {
            return false;
        }
    }

    public void enterEmail(String email) {
        typeWithClear(emailInput, email);
    }

    public void clickLoginButton() {
        click(loginButton);
    }

    public DashboardPage login(String email) {
        enterEmail(email);
        clickLoginButton();
        waitForLoginComplete();
        return new DashboardPage(driver);
    }

    public DashboardPage loginWithDefaultUser() {
        return login(ConfigReader.getTestUserEmail());
    }

    private void waitForLoginComplete() {
        waitUtils.waitForElementInvisible(loadingSpinner);
        waitUtils.waitForUrlContains("/dashboard");
    }

    public boolean isLoginButtonEnabled() {
        return isElementEnabled(loginButton);
    }

    public boolean isLoginButtonDisabled() {
        return !isLoginButtonEnabled();
    }

    public String getInfoAlertText() {
        try {
            return getText(infoAlert);
        } catch (Exception e) {
            return null;
        }
    }

    public String getErrorAlertText() {
        try {
            return getText(errorAlert);
        } catch (Exception e) {
            return null;
        }
    }

    public boolean isErrorAlertDisplayed() {
        return isElementDisplayed(errorAlert);
    }

    public String getEmailInputValue() {
        return getAttribute(emailInput, "value");
    }

    public void clearEmailInput() {
        WebDriver.TargetLocator targetLocator = driver.switchTo();
        driver.findElement(emailInput).clear();
    }
}
